File: web_app/frontend/patient.js
// Patient Frontend Complete Logic
function toggleAuth(view) {
    document.getElementById('login-form').classList.add('hidden');
    document.getElementById('register-form').classList.add('hidden');
    document.getElementById('otp-form').classList.add('hidden');

    if (view === 'login') {
        document.getElementById('login-form').classList.remove('hidden');
        document.getElementById('auth-title').innerText = "🩺 Patient Login";
    } else if (view === 'register') {
        document.getElementById('register-form').classList.remove('hidden');
        document.getElementById('auth-title').innerText = "🩺 Patient Registration";
    } else if (view === 'otp') {
        document.getElementById('otp-form').classList.remove('hidden');
        document.getElementById('auth-title').innerText = "🩺 Email Verification";
    }
}

let pendingEmail = "";
let currentUser = null;

async function doLogin() {
    const userid = document.getElementById('login-userid').value;
    const pwd = document.getElementById('login-pwd').value;
    if (!userid || !pwd) return alert("Please enter User ID and Password.");

    try {
        const res = await fetch('/api/patient/login', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ userid, password: pwd })
        });
        const data = await res.json();
        if (res.ok) {
            currentUser = data.user;
            localStorage.setItem('patientSession', JSON.stringify(currentUser));
            document.getElementById('auth-view').classList.add('hidden');
            document.getElementById('dashboard-view').classList.remove('hidden');
            document.getElementById('dash-name').innerText = currentUser.name;
            loadMedicalHistory();
            loadFamily();
        } else {
            alert(data.detail || "Login failed");
        }
    } catch (e) {
        alert("Error connecting to server.");
    }
}

async function doRegister() {
    const d = {
        name: document.getElementById('reg-name').value,
        dob: document.getElementById('reg-dob').value,
        gender: document.getElementById('reg-gender').value,
        blood: document.getElementById('reg-blood').value,
        contact: document.getElementById('reg-contact').value,
        email: document.getElementById('reg-email').value,
        address: document.getElementById('reg-address').value,
        allergies: document.getElementById('reg-allergies').value,
        userid: document.getElementById('reg-userid').value,
        pwd: document.getElementById('reg-pwd').value,
    };
    const cpwd = document.getElementById('reg-cpwd').value;

    if (!d.name || !d.dob || !d.email || !d.userid || !d.pwd) return alert("Please fill required details");
    if (d.pwd !== cpwd) return alert("Passwords do not match");

    try {
        const res = await fetch('/api/patient/register', {
            method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(d)
        });
        const data = await res.json();
        if (res.ok) {
            pendingEmail = d.email;
            alert("OTP Sent to Email!");
            toggleAuth('otp');
        } else alert(data.detail || "Registration failed");
    } catch (e) { alert("Error connecting to server."); }
}

async function verifyOTP() {
    const otp = document.getElementById('verify-otp').value;
    if (!otp) return alert("Enter OTP");

    try {
        const res = await fetch('/api/patient/verify-otp', {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email: pendingEmail, otp })
        });
        const data = await res.json();
        if (res.ok) {
            alert("Registration Successful!");
            toggleAuth('login');
        } else alert(data.detail || "Invalid OTP");
    } catch (e) { alert("Error connecting to server."); }
}

function switchTab(tabId) {
    ['emergency', 'medical', 'family', 'account'].forEach(t => document.getElementById('tab-' + t).classList.add('hidden'));
    document.querySelectorAll('.tab').forEach(t => t.classList.remove('active'));
    document.getElementById('tab-' + tabId).classList.remove('hidden');
    if (event) event.target.classList.add('active');

    if (tabId === 'account' && currentUser) {
        document.getElementById('acc-contact').value = currentUser.contact_number || '';
        document.getElementById('acc-address').value = currentUser.address || '';
        document.getElementById('acc-allergies').value = currentUser.known_allergies || '';
    }
}

function logout() {
    currentUser = null;
    localStorage.removeItem('patientSession');
    document.getElementById('dashboard-view').classList.add('hidden');
    document.getElementById('auth-view').classList.remove('hidden');
    document.getElementById('login-pwd').value = '';
}

let currentLat = null;
let currentLon = null;

async function searchHospitals() {
    const treatment = document.getElementById('srch-treatment').value;
    const sort_by = document.getElementById('srch-sort').value;
    if (!treatment) return alert("Please enter a treatment");

    const container = document.getElementById('search-results');
    container.innerHTML = "<p>Retrieving your location for accurate distances...</p>";

    if ("geolocation" in navigator) {
        try {
            const pos = await new Promise((resolve, reject) => {
                navigator.geolocation.getCurrentPosition(resolve, reject, {
                    enableHighAccuracy: true,
                    timeout: 10000,
                    maximumAge: 0
                });
            });
            currentLat = pos.coords.latitude;
            currentLon = pos.coords.longitude;
        } catch (err) {
            console.warn("Geolocation blocked or failed. Using default location (Mumbai).");
        }
    }

    container.innerHTML = "<p>Searching hospitals...</p>";

    try {
        const res = await fetch('/api/patient/emergency-search', {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ treatment, sort_by, lat: currentLat, lon: currentLon })
        });
        const data = await res.json();
        if (data.results && data.results.length > 0) {
            let html = '';
            data.results.forEach(h => {
                html += `
                <div class="h-card">
                    <h4 style="color:var(--color-patient)">${h.name}</h4>
                    <p style="font-size:0.9rem; margin:0.5rem 0;">Distance: <strong>${h.distance_km.toFixed(1)} km</strong> | Cost: ₹${h.cost} | Rating: ${h.avg_rating > 0 ? h.avg_rating.toFixed(1) : 'New'}</p>
                    <p style="font-size:0.8rem; color:var(--text-muted); margin-bottom:1rem;">Beds: ${h.available_beds} | Contact: ${h.contact_number} | ${h.address_label}</p>
                    <div style="margin-bottom:0.75rem; display:flex; align-items:center;">
                        <input type="checkbox" id="req-amb-${h.hospital_id}" style="accent-color:var(--color-patient); width:1.2rem; height:1.2rem; margin-right:0.5rem; cursor:pointer;">
                        <label for="req-amb-${h.hospital_id}" style="font-size:0.9rem; cursor:pointer;" title="If checked, we will dispatch an ambulance if one is available.">Request Ambulance Trip</label>
                    </div>
                    <div style="display:flex; gap:0.5rem; margin-top:0.5rem; flex-wrap:wrap;">
                        <button class="btn btn-patient" style="flex:1; min-width:100px; padding: 0.5rem;" onclick="bookBed(${h.hospital_id}, ${h.treatment_id}, '${h.name}')">Book Bed</button>
                        <button class="btn" style="flex:1; min-width:100px; padding: 0.5rem; background:#3b82f6; color:white;" onclick="viewReviews(${h.hospital_id}, '${h.name.replace(/'/g, "\\'")}')">View Reviews</button>
                        <a href="tel:${h.contact_number}" class="btn" style="flex:1; min-width:100px; padding: 0.5rem; text-align:center; background:#10B981; color:white; text-decoration:none;">Call Hospital</a>
                    </div>
                </div>`;
            });
            container.innerHTML = html;
        } else {
            container.innerHTML = `<p style="color:var(--text-muted)">No hospitals found for '${treatment}'.</p>`;
        }
    } catch (e) { container.innerHTML = "<p>Error loading hospitals.</p>"; }
}

async function bookBed(hospital_id, treatment_id) {
    if (!confirm("Confirm bed booking?")) return;
    try {
        const whom = document.getElementById('srch-whom').value;
        const patientName = whom === 'self' ? currentUser.name : whom;

        let reqAmb = false;
        const ambCheck = document.getElementById('req-amb-' + hospital_id);
        if (ambCheck) reqAmb = ambCheck.checked;

        const res = await fetch('/api/patient/book-bed', {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                email: currentUser.email,
                name: patientName,
                hospital_id,
                treatment_id,
                requires_ambulance: reqAmb,
                patient_lat: currentLat,
                patient_lon: currentLon
            })
        });
        const data = await res.json();
        if (res.ok) {
            document.getElementById('bs-hospital-name').innerText = document.getElementById('srch-treatment').value + " at " + (arguments[2] || "Hospital");

            if (data.driver) {
                document.getElementById('bs-no-driver').style.display = 'none';
                document.getElementById('bs-driver-info').style.display = 'block';
                document.getElementById('bs-driver-name').innerText = data.driver.name;
                document.getElementById('bs-driver-call').href = 'tel:' + data.driver.contact_number;
            } else {
                document.getElementById('bs-driver-info').style.display = 'none';
                document.getElementById('bs-no-driver').style.display = 'block';
                document.getElementById('bs-no-driver').innerText = "Currently no ambulances available at this hospital.";
            }

            document.getElementById('booking-success-modal').style.display = 'flex';
            loadMedicalHistory();
        } else alert(data.detail || "Booking failed.");
    } catch (e) { alert("Communication error."); }
}

function closeBookingSuccess() {
    document.getElementById('booking-success-modal').style.display = 'none';
    switchTab('medical');
}

async function loadMedicalHistory() {
    if (!currentUser) return;
    try {
        const res = await fetch('/api/patient/history/' + currentUser.email);
        const data = await res.json();
        const container = document.getElementById('history-results');
        if (data.bookings && data.bookings.length > 0) {
            let html = ``;
            data.bookings.forEach(b => {
                let col = b.status === 'Discharged' ? '#10B981' : '#F59E0B';

                let rateBtnHtml = '';
                if (b.status === 'Discharged') {
                    if (b.is_rated) {
                        rateBtnHtml = `<button class="btn" style="padding:0.25rem 0.5rem; background:#e2e8f0; font-size:0.8rem; margin-top:0.5rem; width:100%;" onclick="viewMyRating('${b.booking_id}', '${b.hospital_name}')">View My Rating</button>`;
                    } else {
                        rateBtnHtml = `<button class="btn btn-patient" style="padding:0.25rem 0.5rem; font-size:0.8rem; margin-top:0.5rem; width:100%;" onclick="openRateModal('${b.booking_id}', ${b.hospital_id}, '${b.hospital_name.replace(/'/g, "\\'")}')">Rate Treatment</button>`;
                    }
                }

                html += `
                <div class="h-card" style="display:flex; justify-content:space-between; align-items:center;">
                    <div><h4 style="color:var(--color-patient)">${b.hospital_name}</h4><p style="font-size:0.9rem; margin-top:0.5rem">Treatment: ${b.treatment_name || 'N/A'}</p></div>
                    <div style="text-align:right;">
                        <span style="font-weight:bold; color:${col}">• ${b.status}</span>
                        <p style="font-size:0.8rem; margin-top:0.25rem;">Id: ${b.booking_id}</p>
                        ${rateBtnHtml}
                    </div>
                </div>`;
            });
            container.innerHTML = html;
        } else container.innerHTML = `<p style="color:var(--text-muted)">No past records found.</p>`;
    } catch (e) { console.error(e); }
}

async function loadFamily() {
    try {
        const res = await fetch('/api/patient/family/' + currentUser.patient_id);
        const data = await res.json();
        const list = document.getElementById('family-results');
        const whomDropdown = document.getElementById('fam-dropdown');
        if (data.family && data.family.length > 0) {
            let html = '';
            let optHtml = '<option value="self">Myself</option>';
            data.family.forEach(f => {
                html += `<div class="h-card" style="width:250px;">
                    <h4 style="margin-bottom:0.5rem;">${f.member_name}</h4>
                    <p style="font-size:0.9rem; color:var(--text-muted)">Relation: ${f.relation}</p>
                    <p style="font-size:0.9rem; color:var(--text-muted)">DOB: ${f.dob}</p>
                    <button class="btn" style="background:#ef4444; margin-top:1rem; padding:0.4rem;" onclick="delFamily(${f.member_id})">Delete</button>
                </div>`;
                optHtml += `<option value="${f.member_name}">${f.member_name} (${f.relation})</option>`;
            });
            list.innerHTML = html;
            whomDropdown.innerHTML = optHtml;
        } else {
            list.innerHTML = `<p style="color:var(--text-muted)">No family members added.</p>`;
            whomDropdown.innerHTML = '';
        }
    } catch (e) { console.error(e); }
}

async function addFamily() {
    const name = document.getElementById('fam-name').value;
    const relation = document.getElementById('fam-rel').value;
    const dob = document.getElementById('fam-dob').value;
    if (!name || !relation || !dob) return alert("Fill all family fields.");
    try {
        const res = await fetch('/api/patient/family', {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ patient_id: currentUser.patient_id, name, relation, dob })
        });
        if (res.ok) {
            alert("Added successfully.");
            loadFamily();
        } else alert("Failed to add.");
    } catch (e) { alert("Error."); }
}

async function delFamily(member_id) {
    if (!confirm("Are you sure?")) return;
    try {
        const res = await fetch('/api/patient/family/' + member_id, { method: 'DELETE' });
        if (res.ok) loadFamily();
    } catch (e) { console.error(e); }
}

async function updateAccount() {
    const cur_pwd = document.getElementById('acc-pwd').value;
    const new_uid = document.getElementById('acc-newuid').value;
    const new_pwd = document.getElementById('acc-newpwd').value;
    const contact = document.getElementById('acc-contact').value;
    const address = document.getElementById('acc-address').value;
    const allergies = document.getElementById('acc-allergies').value;

    if (!cur_pwd) return alert("Current password is required.");
    try {
        const res = await fetch('/api/patient/update-account', {
            method: 'POST', headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ patient_id: currentUser.patient_id, cur_pwd, new_uid, new_pwd, contact, address, allergies })
        });
        const data = await res.json();
        if (res.ok) {
            alert("Account updated successfully.");
            currentUser = data.user;
        } else alert(data.detail || "Update failed.");
    } catch (e) { alert("Error communicating with server."); }
}

async function viewReviews(hospital_id, hospital_name) {
    document.getElementById('rm-title').innerText = `${hospital_name} - Patient Reviews`;
    const c = document.getElementById('rm-content');
    c.innerHTML = "<em>Loading reviews...</em>";
    document.getElementById('reviews-modal').style.display = 'flex';

    try {
        const res = await fetch(`/api/patient/reviews/${hospital_id}`);
        const data = await res.json();
        if (data.reviews && data.reviews.length > 0) {
            c.innerHTML = data.reviews.map(r => `
                <div style="border-bottom:1px solid var(--border); padding-bottom:0.5rem;">
                    <div style="display:flex; justify-content:space-between; margin-bottom:0.25rem;">
                        <strong>Rating: ${r.overall_rating} ★</strong>
                        <span style="font-size:0.8rem; color:var(--text-muted);">${r.date_submitted}</span>
                    </div>
                    <p style="font-size:0.9rem; margin:0; line-height:1.4;">"${r.written_review}"</p>
                </div>
            `).join('');
        } else {
            c.innerHTML = "<p>No written reviews available for this hospital yet.</p>";
        }
    } catch (e) {
        c.innerHTML = "<p style='color:red;'>Failed to load reviews.</p>";
    }
}

async function deletePatientAccount() {
    if (!confirm("Are you entirely sure you want to delete your account? This will erase all medical history.")) return;
    try {
        const res = await fetch(`/api/patient/profile/${currentUser.patient_id}`, { method: 'DELETE' });
        if (res.ok) {
            alert("Account deleted.");
            logout();
        } else alert("Failed to delete account.");
    } catch (e) { }
}

function openRateModal(booking_id, hospital_id, hospital_name) {
    document.getElementById('rate-booking-id').value = booking_id;
    document.getElementById('rate-hospital-id').value = hospital_id;
    document.getElementById('rate-hospital-name').innerText = hospital_name;

    document.querySelectorAll('.star-rating').forEach(group => {
        group.setAttribute('data-value', '5');
        group.querySelectorAll('.star').forEach(s => s.classList.add('active'));
    });

    document.getElementById('rate-text').value = '';

    document.getElementById('rate-modal').style.display = 'flex';
}

async function submitRating() {
    const booking_id = document.getElementById('rate-booking-id').value;
    const hospital_id = parseInt(document.getElementById('rate-hospital-id').value);

    const doctor_care = parseInt(document.querySelector('.star-rating[data-id="rate-doctor"]').getAttribute('data-value'));
    const hygiene = parseInt(document.querySelector('.star-rating[data-id="rate-hygiene"]').getAttribute('data-value'));
    const staff_behavior = parseInt(document.querySelector('.star-rating[data-id="rate-staff"]').getAttribute('data-value'));
    const facilities = parseInt(document.querySelector('.star-rating[data-id="rate-facilities"]').getAttribute('data-value'));
    const speed_of_service = parseInt(document.querySelector('.star-rating[data-id="rate-speed"]').getAttribute('data-value'));

    const review_text = document.getElementById('rate-text').value;

    document.getElementById('rate-submit-btn').disabled = true;
    document.getElementById('rate-submit-btn').innerText = "Submitting...";

    try {
        const res = await fetch('/api/patient/rate', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ booking_id, hospital_id, doctor_care, hygiene, staff_behavior, facilities, speed_of_service, review_text })
        });

        if (res.ok) {
            alert("Thank you for your rating!");
            document.getElementById('rate-modal').style.display = 'none';
            loadMedicalHistory(); // refresh to show 'View My Rating'
        } else {
            const data = await res.json();
            alert(data.detail || "Error submitting rating.");
        }
    } catch (e) {
        alert("Serverside error.");
    } finally {
        document.getElementById('rate-submit-btn').disabled = false;
        document.getElementById('rate-submit-btn').innerText = "Submit Rating";
    }
}

async function viewMyRating(booking_id, hospital_name) {
    document.getElementById('rm-title').innerText = `Your Rating for ${hospital_name}`;
    const c = document.getElementById('rm-content');
    c.innerHTML = "<em>Loading your rating...</em>";
    document.getElementById('reviews-modal').style.display = 'flex';

    try {
        const res = await fetch(`/api/patient/my-review/${booking_id}`);
        if (res.ok) {
            const data = await res.json();
            const r = data.review;
            c.innerHTML = `
                <div style="padding:1rem; background: rgba(0, 119, 182, 0.05); border-radius: 8px;">
                    <div style="font-size:1.2rem; font-weight:bold; color:var(--color-patient); margin-bottom:1rem; text-align:center;">
                        Overall Rating: ${r.overall_rating.toFixed(1)} ★
                    </div>
                    <div style="display:flex; flex-wrap:wrap; gap:1rem; font-size:0.9rem; margin-bottom:1rem; justify-content:center;">
                        <span style="background:#fff; padding:0.25rem 0.5rem; border-radius:4px; border:1px solid #ccc;">Doctor's Care: ${r.doctor_care}</span>
                        <span style="background:#fff; padding:0.25rem 0.5rem; border-radius:4px; border:1px solid #ccc;">Hygiene: ${r.hygiene}</span>
                        <span style="background:#fff; padding:0.25rem 0.5rem; border-radius:4px; border:1px solid #ccc;">Staff Behavior: ${r.staff_behavior}</span>
                        <span style="background:#fff; padding:0.25rem 0.5rem; border-radius:4px; border:1px solid #ccc;">Facilities: ${r.facilities}</span>
                        <span style="background:#fff; padding:0.25rem 0.5rem; border-radius:4px; border:1px solid #ccc;">Speed: ${r.speed_of_service}</span>
                    </div>
                    <p style="font-size:0.95rem; line-height:1.5;">"${r.written_review || "No written feedback provided."}"</p>
                </div>
            `;
        } else {
            c.innerHTML = "<p style='color:red;'>Could not fetch your rating.</p>";
        }
    } catch (e) {
        c.innerHTML = "<p style='color:red;'>Error connecting to server.</p>";
    }
}

// Initialize Star Ratings
document.addEventListener('DOMContentLoaded', () => {
    const saved = localStorage.getItem('patientSession');
    if (saved) {
        currentUser = JSON.parse(saved);
        document.getElementById('auth-view').classList.add('hidden');
        document.getElementById('dashboard-view').classList.remove('hidden');
        document.getElementById('dash-name').innerText = currentUser.name;
        switchTab('search');
    }
    document.querySelectorAll('.star-rating').forEach(group => {
        const stars = group.querySelectorAll('.star');
        stars.forEach(star => {
            star.addEventListener('click', (e) => {
                const val = parseInt(e.target.getAttribute('data-val'));
                group.setAttribute('data-value', val);
                stars.forEach(s => {
                    const sVal = parseInt(s.getAttribute('data-val'));
                    if (sVal <= val) {
                        s.classList.add('active');
                    } else {
                        s.classList.remove('active');
                    }
                });
            });
        });
    });
});

File: web_app/backend/routers/patient.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from database import get_db_connection
import mysql.connector
import datetime
import random
import bcrypt

def send_otp(email, name, otp):
    try:
        try:
            from utils import send_registration_otp_email
        except ImportError:
            from backend.utils import send_registration_otp_email
        success = send_registration_otp_email(email, name, otp)
        if not success:
            raise Exception("send_registration_otp_email returned False")
    except Exception as e:
        raise Exception(f"SMTP FAILED: {str(e)}")

router = APIRouter()

class PatientLoginRequest(BaseModel):
    userid: str
    password: str

class PatientRegisterRequest(BaseModel):
    name: str # maps to full_name
    dob: str
    gender: str
    blood: str
    contact: str
    email: str
    address: str
    allergies: str
    userid: str
    pwd: str

class OTPVerifyRequest(BaseModel):
    email: str
    otp: str

class SearchRequest(BaseModel):
    treatment: str
    sort_by: str
    lat: Optional[float] = None
    lon: Optional[float] = None

class BookBedRequest(BaseModel):
    email: str
    name: str
    hospital_id: int
    treatment_id: int

class FamilyAddRequest(BaseModel):
    patient_id: int
    name: str # maps to full_name
    relation: str
    dob: str

class AccountUpdateRequest(BaseModel):
    patient_id: int
    cur_pwd: str
    new_uid: Optional[str] = None
    new_pwd: Optional[str] = None
    contact: Optional[str] = None
    address: Optional[str] = None
    allergies: Optional[str] = None

temp_regs = {}

@router.post("/login")
def login(req: PatientLoginRequest):
    try:
        conn = get_db_connection()
        cur = conn.cursor(dictionary=True)
        cur.execute("SELECT * FROM patients WHERE userid=%s", (req.userid,))
        user = cur.fetchone()
        conn.close()
        
        has_access = False
        if user:
            db_pwd = user['password'].encode('utf-8')
            req_pwd = req.password.encode('utf-8')
            if db_pwd.startswith(b'$2b$'):
                has_access = bcrypt.checkpw(req_pwd, db_pwd)
            else:
                has_access = (user['password'] == req.password)
                
        if has_access:
            user['name'] = user['full_name']
            return {"user": user, "message": "Login successful"}
        else:
            raise HTTPException(status_code=401, detail="Invalid User ID or Password.")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/register")
def register(req: PatientRegisterRequest):
    try:
        conn = get_db_connection()
        cur = conn.cursor()
        
        # Check global uniqueness of User ID
        try:
            from utils import is_userid_unique
        except ImportError:
            from backend.utils import is_userid_unique
            
        if not is_userid_unique(cur, req.userid):
            conn.close()
            raise HTTPException(status_code=400, detail="User ID is already taken by another account.")
            
        cur.execute("SELECT patient_id FROM patients WHERE email=%s", (req.email,))
        if cur.fetchone():
            conn.close()
            raise HTTPException(status_code=400, detail="Email already exists.")
        conn.close()
        
        otp = str(random.randint(100000, 999999))
        temp_regs[req.email] = {"data": req.dict(), "otp": otp}
        
        try:
            send_otp(req.email, req.name, otp)
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"SMTP FAILED: {str(e)}")
            
        return {"message": "OTP sent"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/verify-otp")
def verify_otp(req: OTPVerifyRequest):
    if req.email not in temp_regs or temp_regs[req.email]["otp"] != req.otp:
        raise HTTPException(status_code=400, detail="Invalid OTP")
    
    d = temp_regs[req.email]["data"]
    try:
        conn = get_db_connection()
        cur = conn.cursor()
        query = """INSERT INTO patients 
                   (full_name, dob, gender, blood_group, contact_number, email, address, known_allergies, userid, password) 
                   VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)"""
        hashed = bcrypt.hashpw(d['pwd'].encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
        cur.execute(query, (d['name'], d['dob'], d['gender'], d['blood'], d['contact'], d['email'], d['address'], d['allergies'], d['userid'], hashed))
        conn.commit()
        conn.close()
        del temp_regs[req.email]
        return {"message": "Registration complete"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/emergency-search")
def emergency_search(req: SearchRequest):
    try:
        conn = get_db_connection()
        cur = conn.cursor(dictionary=True)
        
        # Use provided geolocation if allowed by browser, else fallback
        patient_lat = req.lat if req.lat is not None else 19.0760
        patient_lon = req.lon if req.lon is not None else 72.8777
        
        # Fixed ratings -> reviews and rating -> overall_rating
        query = """SELECT h.hospital_id, h.name, h.address_label, h.contact_number, h.latitude, h.longitude,
                   t.treatment_id, t.cost, t.available_beds,
                   IFNULL((SELECT AVG(overall_rating) FROM reviews WHERE hospital_id=h.hospital_id), 0) as avg_rating
                   FROM hospitals h
                   JOIN treatments t ON h.hospital_id = t.hospital_id
                   WHERE t.treatment_name = %s AND t.available_beds > 0"""
        cur.execute(query, (req.treatment,))
        results = cur.fetchall()
        conn.close()
        
        final_results = []
        for r in results:
            dist = abs(float(r['latitude']) - patient_lat) * 111.0 + abs(float(r['longitude']) - patient_lon) * 111.0
            r['distance_km'] = dist
            final_results.append(r)
            
        if "Nearest" in req.sort_by:
            final_results.sort(key=lambda x: x['distance_km'])
        elif "Cost" in req.sort_by:
            final_results.sort(key=lambda x: x['cost'])
        elif "Rating" in req.sort_by:
            final_results.sort(key=lambda x: x['avg_rating'], reverse=True)
            
        return {"results": final_results[:10]}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/reviews/{hospital_id}")
def get_hospital_reviews(hospital_id: int):
    try:
        conn = get_db_connection()
        cur = conn.cursor(dictionary=True)
        query = """SELECT r.overall_rating, r.written_review, r.date_submitted,
                   (r.doctor_care + r.hygiene + r.staff_behavior + r.facilities + r.speed_of_service) / 5 as avg_cat
                   FROM reviews r
                   WHERE r.hospital_id = %s AND r.written_review IS NOT NULL AND r.written_review != ''
                   ORDER BY r.date_submitted DESC LIMIT 10"""
        cur.execute(query, (hospital_id,))
        reviews = cur.fetchall()
        conn.close()
        return {"reviews": reviews}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/book-bed")
def book_bed(req: BookBedRequest):
    try:
        import datetime
        conn = get_db_connection()
        cur = conn.cursor(dictionary=True)
        
        now = datetime.datetime.now()
        clean_name = "".join([c for c in req.name if c.isalpha()]).lower()
        booking_id = f"{clean_name}{now.strftime('%y%m%d%H%M%S')}"

        query_b = "INSERT INTO bookings (booking_id, patient_email, patient_name, hospital_id, treatment_id, status) VALUES (%s, %s, %s, %s, %s, %s)"
        cur.execute(query_b, (booking_id, req.email, req.name, req.hospital_id, req.treatment_id, "Admitted"))
        
        query_u = "UPDATE treatments SET available_beds = available_beds - 1 WHERE treatment_id = %s"
        cur.execute(query_u, (req.treatment_id,))
        
        # Find available driver
        cur.execute("SELECT driver_id, full_name as name, contact_number, vehicle_type, vehicle_number FROM drivers WHERE affiliated_hospital_id=%s AND status='Available' LIMIT 1", (req.hospital_id,))
        driver = cur.fetchone()
        
        if driver:
            cur.execute("UPDATE drivers SET status='Busy' WHERE driver_id=%s", (driver['driver_id'],))
        
        conn.commit()
        conn.close()
        
        msg = f"EMERGENCY: Patient {req.name} successfully booked a bed."
        wa_link = f"https://wa.me/919999999999?text={msg.replace(' ', '%20')}"
        return {"message": "Booked", "wa_link": wa_link, "driver": driver}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/history/{email}")
def get_history(email: str):
    try:
        conn = get_db_connection()
        cur = conn.cursor(dictionary=True)
        query = """SELECT b.booking_id, b.status, h.name as hospital_name, t.treatment_name 
                   FROM bookings b
                   JOIN hospitals h ON b.hospital_id = h.hospital_id
                   LEFT JOIN treatments t ON b.treatment_id = t.treatment_id
                   WHERE b.patient_email = %s ORDER BY b.booking_id DESC"""
        cur.execute(query, (email,))
        records = cur.fetchall()
        conn.close()
        return {"bookings": records}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/family/{patient_id}")
def get_family(patient_id: int):
    try:
        conn = get_db_connection()
        cur = conn.cursor(dictionary=True)
        cur.execute("SELECT member_id, patient_id, full_name, relation, dob FROM family_members WHERE patient_id=%s", (patient_id,))
        records = cur.fetchall()
        if records:
            for r in records:
                if isinstance(r['dob'], datetime.date):
                    r['dob'] = r['dob'].isoformat()
                r['member_name'] = r['full_name']
        conn.close()
        return {"family": records}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/family")
def add_family(req: FamilyAddRequest):
    try:
        conn = get_db_connection()
        cur = conn.cursor()
        cur.execute("INSERT INTO family_members (patient_id, full_name, relation, dob) VALUES (%s, %s, %s, %s)", 
                    (req.patient_id, req.name, req.relation, req.dob))
        conn.commit()
        conn.close()
        return {"message": "Success"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/family/{member_id}")
def delete_family(member_id: int):
    try:
        conn = get_db_connection()
        cur = conn.cursor()
        cur.execute("DELETE FROM family_members WHERE member_id=%s", (member_id,))
        conn.commit()
        conn.close()
        return {"message": "Success"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/update-account")
def update_account(req: AccountUpdateRequest):
    try:
        conn = get_db_connection()
        cur = conn.cursor(dictionary=True)
        cur.execute("SELECT * FROM patients WHERE patient_id=%s AND password=%s", (req.patient_id, req.cur_pwd))
        user = cur.fetchone()
        if not user:
            conn.close()
            raise HTTPException(status_code=400, detail="Incorrect current password")
            
        uid = req.new_uid if req.new_uid else user['userid']
        pwd = req.new_pwd if req.new_pwd else user['password']
        
        if req.new_uid and req.new_uid != user['userid']:
            try:
                from utils import is_userid_unique
            except ImportError:
                from backend.utils import is_userid_unique
            if not is_userid_unique(cur, req.new_uid):
                conn.close()
                raise HTTPException(status_code=400, detail="User ID is already taken by another account")
                
        if req.new_pwd:
            pwd = bcrypt.hashpw(req.new_pwd.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
            
        if req.contact:
            cur.execute("UPDATE patients SET contact_number=%s WHERE patient_id=%s", (req.contact, req.patient_id))
        if req.address:
            cur.execute("UPDATE patients SET address=%s WHERE patient_id=%s", (req.address, req.patient_id))
        if req.allergies:
            cur.execute("UPDATE patients SET known_allergies=%s WHERE patient_id=%s", (req.allergies, req.patient_id))
            
        cur.execute("UPDATE patients SET userid=%s, password=%s WHERE patient_id=%s", (uid, pwd, req.patient_id))
        conn.commit()
        
        cur.execute("SELECT * FROM patients WHERE patient_id=%s", (req.patient_id,))
        user = cur.fetchone()
        user['name'] = user['full_name']
        conn.close()
        return {"message": "Success", "user": user}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/profile/{patient_id}")
def delete_profile(patient_id: int):
    try:
        conn = get_db_connection()
        cur = conn.cursor()
        cur.execute("DELETE FROM patients WHERE patient_id=%s", (patient_id,))
        conn.commit()
        conn.close()
        return {"message": "Success"}
    except Exception as e:
        try:
            cur.execute("DELETE FROM family_members WHERE patient_id=%s", (patient_id,))
            cur.execute("DELETE FROM patients WHERE patient_id=%s", (patient_id,))
            conn.commit()
        except:
            pass
        finally:
            conn.close()
        return {"message": "Attempted to delete"}
